zenith_utils/src/lsn.rs

```diff
@@ -0,0 +1,138 @@
+// Doing so would mean zenith_utils gets a dependency on postgres-types.
+// That's not necessarily a bad thing, but if there is only one place
+// that needs it, it might be easier to do it there, e.g.
+//     let x = Lsn::from(u64::from(pglsn))
+
+/// A Postgres LSN (Log Sequence Number), also known as an XLogRecPtr
+#[derive(Debug, Clone, Copy, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Lsn(pub u64);
+
+/// We tried to parse an LSN from a string, but failed
+#[derive(Debug, PartialEq, thiserror::Error)]
+#[error("LsnParseError")]
+pub struct LsnParseError;
+
+impl Lsn {
+    /// Maximum possible value for an LSN
+    pub const MAX: Lsn = Lsn(u64::MAX);
+
+    /// Parse an LSN from a filename in the form `0000000000000000`
+    pub fn from_filename<F>(filename: F) -> Result<Self, LsnParseError>
+    where
+        F: AsRef<Path>,
+    {
+        let filename: &Path = filename.as_ref();
+        let filename = filename.to_str().ok_or(LsnParseError)?;
+        let n = u64::from_str_radix(filename, 16).or(Err(LsnParseError))?;
+        Ok(Lsn(n))
+    }
+
+    // FIXME: Is there a more ergonomic way to do this?
+    //
+    // For example, we could have a SegmentGeometry struct that
+    // implements this logic instead, e.g.
+    //
+    //    const seg_geo = SegmentGeometry::new(16 * 1024 * 1024);
+    //    let offset = seg_geo.seg_offset(lsn);
+    //    let seg_no = seg_geo.seg_num(lsn);
+
+    /// Compute the offset into a segment
+    pub fn segment_offset(self, seg_sz: u64) -> u64 {
+        self.0 % seg_sz
+    }
+
+    /// Compute the segment number
+    // FIXME: should segment number be a real type?
+    pub fn segment_number(self, seg_sz: u64) -> u64 {
+        self.0 / seg_sz
+    }
+}
+
+impl From<u64> for Lsn {
+    fn from(n: u64) -> Self {
+        Lsn(n)
+    }
+}
+
+impl From<Lsn> for u64 {
+    fn from(lsn: Lsn) -> u64 {
+        lsn.0
+    }
+}
+
+impl FromStr for Lsn {
+    type Err = LsnParseError;
+
+    /// Parse an LSN from a string in the form `00000000/00000000`
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut splitter = s.split('/');
+        if let (Some(left), Some(right), None) = (splitter.next(), splitter.next(), splitter.next())
+        {
+            let left_num = u32::from_str_radix(left, 16).map_err(|_| LsnParseError)?;
+            let right_num = u32::from_str_radix(right, 16).map_err(|_| LsnParseError)?;
+            Ok(Lsn((left_num as u64) << 32 | right_num as u64))
+        } else {
+            Err(LsnParseError)
+        }
+    }
+}
+
+impl fmt::Display for Lsn {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // FIXME: should this be "{:X}/{:08X}" ?
+        write!(f, "{:X}/{:X}", self.0 >> 32, self.0 & 0xffffffff)
+    }
+}
+
+// FIXME: add a LsnDelta type?
+impl Sub<u64> for Lsn {
+    type Output = Lsn;
+
+    fn sub(self, other: u64) -> Self::Output {
+        // panic if the subtraction overflows/underflows.
```

Comment:
We could also drop `sub` and instead implement `checked_sub`, returning `Option<Lsn>`, forcing the caller to choose a strategy.

In the standard library, there is also [`saturating_sub`](https://doc.rust-lang.org/std/primitive.u64.html#method.saturating_sub), which returns the min or max instead of overflowing.